File: internal/app/deeplink_test.go
// SPDX-License-Identifier: MIT
package app_test

import (
	"testing"

	"renbrowser/internal/app"
	"renbrowser/internal/deeplink"
)

func TestHandleDeepLinkAcceptsAndQueues(t *testing.T) {
	deeplink.ClearPending()
	defer deeplink.ClearPending()

	svc := &app.BrowserService{}
	got := svc.HandleDeepLink("renbrowser:about")
	if got != "about:" {
		t.Fatalf("HandleDeepLink = %q", got)
	}
	if pending := svc.TakePendingDeepLink(); pending != "about:" {
		t.Fatalf("TakePendingDeepLink = %q", pending)
	}
	if again := svc.TakePendingDeepLink(); again != "" {
		t.Fatalf("second take = %q", again)
	}
}

func TestHandleDeepLinkRejectsExternal(t *testing.T) {
	deeplink.ClearPending()
	defer deeplink.ClearPending()

	svc := &app.BrowserService{}
	if got := svc.HandleDeepLink("https://example.com"); got != "" {
		t.Fatalf("expected empty, got %q", got)
	}
	if pending := svc.TakePendingDeepLink(); pending != "" {
		t.Fatalf("pending = %q", pending)
	}
}

func TestHandleDeepLinkMeshAndRNS(t *testing.T) {
	deeplink.ClearPending()
	defer deeplink.ClearPending()

	hash := "abb3ebcd03cb2388a838e70c001291f9"
	svc := &app.BrowserService{}
	cases := []struct {
		in   string
		want string
	}{
		{"renbrowser://" + hash + "/page/x.mu", hash + ":/page/x.mu"},
		{"rns://" + hash + "/page/home.mu", "rns://" + hash + "/page/home.mu"},
		{"renbrowser://open?url=" + hash, hash + ":/page/index.mu"},
	}
	for _, tc := range cases {
		deeplink.ClearPending()
		got := svc.HandleDeepLink(tc.in)
		if got != tc.want {
			t.Fatalf("HandleDeepLink(%q) = %q, want %q", tc.in, got, tc.want)
		}
	}
}

File: frontend/src/lib/browser/deeplink.test.ts
// SPDX-License-Identifier: MIT
import { describe, expect, it } from "vitest";
import { unwrapDeepLink } from "./deeplink";

const hash = "abb3ebcd03cb2388a838e70c001291f9";

describe("unwrapDeepLink", () => {
  it("unwraps renbrowser wrappers", () => {
    expect(unwrapDeepLink("renbrowser:about")).toBe("about:");
    expect(unwrapDeepLink("renbrowser://about")).toBe("about:");
    expect(unwrapDeepLink("renbrowser://open?url=about%3A")).toBe("about:");
    expect(unwrapDeepLink(`renbrowser://${hash}/page/index.mu`)).toBe(`${hash}:/page/index.mu`);
    expect(unwrapDeepLink(`renbrowser://rns/${hash}/page/home.mu`)).toBe(
      `rns://${hash}/page/home.mu`,
    );
  });

  it("passes through internal urls", () => {
    expect(unwrapDeepLink("about:")).toBe("about:");
    expect(unwrapDeepLink(`rns://${hash}/page/x.mu`)).toBe(`rns://${hash}/page/x.mu`);
    expect(unwrapDeepLink(`${hash}:/page/index.mu`)).toBe(`${hash}:/page/index.mu`);
  });

  it("rejects external and dangerous schemes", () => {
    expect(unwrapDeepLink("https://example.com")).toBe("");
    expect(unwrapDeepLink("javascript:alert(1)")).toBe("");
    expect(unwrapDeepLink("renbrowser://open?url=https%3A%2F%2Fevil.test")).toBe("");
    expect(unwrapDeepLink("data:text/html,hi")).toBe("");
    expect(unwrapDeepLink("file:///etc/passwd")).toBe("");
  });

  it("handles edge cases", () => {
    expect(unwrapDeepLink("")).toBe("");
    expect(unwrapDeepLink("   ")).toBe("");
    expect(unwrapDeepLink("renbrowser:")).toBe("");
    expect(unwrapDeepLink("renbrowser://open?url=")).toBe("");
    expect(unwrapDeepLink("about:\0x")).toBe("");
    expect(unwrapDeepLink("  RenBrowser://License  ")).toBe("license:");
  });
});
